### ISSUE TITLE
Changes to bottles cause Homebrew to do unnecessary software reinstallation

### ISSUE THREAD

**Issue by rhaglennydd** [Homebrew/brew#1170]
I've noticed recently that when I look at the sources of some of the updated formulae, the only changes are removal of Mavericks bottles and addition of Sierra bottles. The version number of the actual underlying software has not been updated. For example, I've received updates to php54, php55, php56 and php70 that don't include updates to those versions of PHP (indeed, 5.4 and 5.5 have reached the end of their support cycles). Nevertheless, when I do `brew upgrade` Homebrew insists on reinstalling these PHP versions, which I had compiled with options so it takes awhile. (Not to mention I have to unlink each PHP formula before I can upgrade the next one, which is another issue.)

My issues with this are A) I don't have Sierra or Mavericks, so these updates don't change anything for me; B) Even if I did have Sierra or Mavericks, the addition or removal of a precompiled bottle is irrelevant to a formula that has been compiled from source. Adding/removing a bottle should not trigger a recompilation, that's silly.

**Comment by zmwangx** [comment]
Sorry, but adding/removing a bottle per se does not trigger a recompilation. This is true for almost all core formulae that haven't got version updated but have recently received Sierra bottles, and in such cases, we don't even drop the Mavericks bottle. Your PHPs were recompiled due to revision bumps, e.g. https://github.com/Homebrew/homebrew-php/commit/c783ac4ead929bcc6c86efc0ed881e29776867c2. Now why did they do https://github.com/Homebrew/homebrew-php/pull/3622 and such over at the php tap, I'm not sure. I suggest you open a ticket there to ask them, and have me CC'ed.

**Comment by javian** [comment]
@rhaglennydd Will, it is true that part of the revision bump for the formulas was related to the release of macOS Sierra and that the effect is you described. This is however not the only issue that has been occurring in the php tap which lead to the revision of the bottles. As an example brew migrated to a new directory structure recently and we discovered that a hardcoded link to the old structure was located in the phpize script in all php versions so that we were no longer able to compile extensions and bottle them. This prevented us from providing Sierra support and meant that the bottles needed to be rebuilt after the underlying issue with the phpize script had been resolved.

As for the issues you bring up I hear you but there are a lot of people that upgrade to Sierra so I believe it is important for us to provide bottles also for the new version of the operating system near to its release. Your second issue I believe is hard to change since it part of the design of brew. Some Formulas have external dependencies to other Formulas and the the dependency is updated the bottle has to be rebuilt or the Formula will be broken. It is not so much that it a bottled is added or removed but the revision that is changed which triggers the reinstall/compile to happen.

So in short this is not a solution to your issues but if you have further issues, ideas or code suggestions feel free to send a PR our way (in the php tap) and we can take it from there.

**Comment by zmwangx** [comment]
@javian FYI you don't need a revision bump to bottle for Sierra. Just specify `--keep-old` to `test-bot` and run the job on Homebrew Sierra Testing.

Hardcoded paths are indeed reason for revision though.

**Comment by zmwangx** [comment]
To be more concrete, check out my bottling jobs on https://bot.brew.sh/view/Testing/job/Homebrew%20Sierra%20Testing/ and the [recent commit history in core](https://github.com/Homebrew/homebrew-core/commits) (I'm mass bottling for Sierra right now).

**Comment by javian** [comment]
@zmwangx I don't have access to `test-bot` afaik so the only way I know how to handle it is through revisions.

**Comment by MikeMcQuaid** [comment]
@javian Ping me in Slack and I'll get you access.
